// vscode-chrome-debug-core/out/src/transformers/basePathTransformer.js
/*---------------------------------------------------------
 * Copyright (C) Microsoft Corporation. All rights reserved.
 *--------------------------------------------------------*/
"use strict";
/**
 * Converts a local path from Code to a path on the target.
 */
var BasePathTransformer = (function () {
    function BasePathTransformer() {
    }
    BasePathTransformer.prototype.launch = function (args) {
        return Promise.resolve();
    };
    BasePathTransformer.prototype.attach = function (args) {
        return Promise.resolve();
    };
    BasePathTransformer.prototype.setBreakpoints = function (args) {
        return true;
    };
    BasePathTransformer.prototype.clearTargetContext = function () {
    };
    BasePathTransformer.prototype.scriptParsed = function (scriptPath) {
        return scriptPath;
    };
    BasePathTransformer.prototype.breakpointResolved = function (bp, scriptPath) {
        return scriptPath;
    };
    BasePathTransformer.prototype.stackTraceResponse = function (response) {
    };
    BasePathTransformer.prototype.getTargetPathFromClientPath = function (clientPath) {
        return clientPath;
    };
    BasePathTransformer.prototype.getClientPathFromTargetPath = function (targetPath) {
        return targetPath;
    };
    return BasePathTransformer;
}());
exports.BasePathTransformer = BasePathTransformer;
